### ISSUE TITLE
Return failure response code when the container stops with non-zero exit code

### ISSUE THREAD

**Issue by ScottG489** [ScottG489/docker-ci-prototype#2]
We currently ignore the exit code of the container run by the server. This means that if the build in the container fails then it will return a 200 response. This is not desirable because then the requester can't easily determine if there was a failure or not and can't act appropriately to a failed build.

We should detect the exit code and response accordingly. We should consider what would be an appropriate exit code. This would likely be a 4xx or 5xx.

Here are some arguments for each:

**4xx:**
- If the user provided bad input to the build such as bad credentials or just a generally bad envar then it would be proper to return a 4xx (likely 400). However, this may be difficult to detect. Each build is very unique.
- The build container itself (not just the name) is an input to the system. This means that if it failed, it's still because the user supplied an image for the server to run which had flaws in it. This means the build server itself was successful, but the user provided container was misconfigured and failed. In other words, the user provided bad input (the container itself).

**5xx:**
- Due to the inability to determine what went wrong in the job because each build is unique, this may be the best blanked approach.
- We have more control over errors on the server side (such as when a container exists so we can't even start it). This would mean that if we returned a 500 it was **likely** due to a failure with the build itself. However, it's always possible that the server will have unforeseen errors and could still return a 500.

That said, the important thing is that we don't return a successful response code so at least builds will report as having failed. I think it's also ok for there to be some overlap between failures returned from the server and failures returned from the container. They don't need to be mutually exclusive.

**Note:** that if the container itself fails to run for some reason such as the container not existing we currently return a 500. We may want to revisit this and return a different error code instead. For instance in the example above of a container not existing we should return a 404.

**Comment by ScottG489** [comment]
A few ideas I've settled on

- Don't return a stream. Why is it important for the user to see the text in real time anyways? All they really need to know is if it succeeded or failed. Anything else is superfluous.
- Return the exit code at the end of the stream. This has 2 side effects:
  - Requires some kind of standard interface for the response (i.e. The build output is all the returned text except for the last line which is the status code they need to parse.
  - Requires the user to manually parse the response and determine if the job passed or failed
- Separate response required (very undesirable since it's two requests and makes the server stateful

**Comment by ScottG489** [comment]
I think I will go with idea #1.

It is the simplest solution to implement. It also simplifies the response to not being a stream which I think is not necessarily a requirement. It may be somewhat expected as a feature of a CI server but I think that could be a bit of a deep seated idea/antipattern that you need to watch your CI jobs output as it runs. Usually the only reason you are watching a build is if you are worried about it failing or there is some problem. Since all of these builds can be run all the same locally using the docker container then it should be very infrequent that you need to troubleshoot build issues on the build server itself rather than running them locally and troubleshooting that way. Plus it's not like you won't be getting your job output; it will just have to wait until the job finishes.

I think eventually it could be a feature to the server that can be specified with a flag set in the request. However, I think that will look something like option #2 which also requires changing the format of the response likely to a json object with the log output and the exit code. The complicated part about this is figuring out how to augment the output stream.

**Comment by ScottG489** [comment]
I'd like to point out that this ticket evolved from the original purpose. It was initially intended for only figuring out what error codes to return in the header. Then when it was realized that we had no control over the header when streaming a response we needed to figure out how to solve that problem first. The second issue has been solved by the commit and explained by the last comment.

Next is to solve the problem of what http status code to return. After thinking about it during work to fix the other issue, I think that a 400 is a good code to return. The biggest issue here is that it doesn't necessarily identify the issue as something with the build itself (e.g. if the user provided a bad image name). However, I don't think we necessarily need an error code to signify where the problem happened between the server or container. So maybe that really isn't that big of a deal anyways.

For now I will at least go with returning a 400 when the build itself returns a non-zero exit code.
